src/main/resources/include/hail/Region.h

```diff
@@ -0,0 +1,121 @@
+#ifndef HAIL_REGION_H
+#define HAIL_REGION_H 1
+
+#include "hail/NativeObj.h"
+#include "hail/NativePtr.h"
+#include <cstdlib>
+#include <cstring>
+#include <algorithm>
+#include <memory>
+#include <vector>
+
+namespace hail {
+
+class Region;
+using RegionPtr = std::shared_ptr<Region>;
+
+class Region : public NativeObj {
+private:
+  static constexpr ssize_t kChunkCap = 64*1024;
+  static constexpr ssize_t kMaxSmall = 4*1024;
+  static constexpr ssize_t kNumBigToKeep = 4;
+  
+  struct CharArray {
+    char buf_[1];
+  };
+  
+  class UniqueChunk {
+   public:
+    std::unique_ptr<CharArray> ptr_;
+    ssize_t size_;
+    
+    UniqueChunk() : ptr_(), size_(0) { }
```

Comment:
I think this code has some unnecessary complexity that can be eliminated if we make Region allocate a block by default and we don't allow the chunk to be empty.  I think this should be deleted, and we should allocate in the ssize_t constructor.

Reply:
Or you can not allocate any chunks and have the index point one past the end of the chunk array, you still don't need empty chunks.

Reply:
Now it never has an empty chunk - but it turns out we still a no-argument Chunk constructor to be 
able to use resize() on a std::vector<Chunk>.  We're only ever shrinking the vector, but
resize() doesn't know that.